cookbook/upgrade/bundles.rst

```diff
@@ -0,0 +1,155 @@
+.. index::
+    single: Upgrading; Bundle; Major Version
+
+Upgrading a Third-Party Bundle for a Major Symfony Version
+==========================================================
+
+Symfony 3 was released on November 2015. Although this version doesn't contain
+any new feature, it removes all the backwards compatibility layers included in
+the previous 2.8 version. If your bundle uses any deprecated feature and it's
+published as a third-party bundle, applications upgrading to Symfony 3 will no
+longer be able to use it.
+
+Allow to Install Symfony 3 Components
+-------------------------------------
+
+Most third-party bundles define their Symfony dependencies using the ``~2.N`` or
+``^2.N`` constraints in the ``composer.json`` file. For example:
+
+.. code-block:: json
+
+    {
+        "require": {
+            "symfony/framework-bundle": "~2.3",
+            "symfony/finder": "~2.3",
+            "symfony/validator": "~2.3"
+        }
+    }
+
+These constraints prevent the bundle from using Symfony 3 components, so it makes
+it impossible to install it in a Symfony 3 based application. This issue is very
+easy to solve thanks to the flexibility of Composer dependencies constraints.
+Just replace ``~2.N`` by ``~2.N|~3.0`` (or ``^2.N`` by ``^2.N|~3.0``).
+
+The above example can be updated to work with Symfony 3 as follows:
+
+.. code-block:: json
+
+    {
+        "require": {
+            "symfony/framework-bundle": "~2.3|~3.0",
+            "symfony/finder": "~2.3|~3.0",
+            "symfony/validator": "~2.3|~3.0"
+        }
+    }
+
+Look for Deprecations and Fix Them
+----------------------------------
+
+Besides allowing to install Symfony 3 component, your bundle must stop using
+any feature deprecated in 2.8 version, because they'll throw exceptions in 3.0
+version. The easiest way to detect deprecations is to install the `PHPUnit Bridge`_
+component and then run the test suite.
+
+First, install the component as a ``dev`` dependency of your bundle:
+
+.. code-block:: bash
+
+    $ composer require --dev "symfony/phpunit-bridge"
+
+Then, run your test suite and look for the deprecation list displayed after the
+PHPUnit test report:
+
+.. code-block:: bash
+
+    $ phpunit
+
+    // ... PHPUnit output ...
+
+    Remaining deprecation notices (3)
+
+    The "pattern" option in file ... is deprecated since version 2.2 and will be
+    removed in 3.0. Use the "path" option in the route definition instead ...
+
+    Twig Function "form_enctype" is deprecated. Use "form_start" instead in ...
+
+    The Symfony\Component\Security\Core\SecurityContext class is deprecated since
+    version 2.6 and will be removed in 3.0. Use ...
```

Comment:
this is not the output you get from the PHPunit bridge, you'll get something like:
```
The "_method" requirement is deprecated since version 2.2 and will be removed in 3.0. Use the setMethods() method instead: 36x
    18x in PageAdminTest::testPageShow from Symfony\Cmf\Bundle\SimpleCmsBundle\Tests\WebTest\Admin
    18x in PageAdminTest::testPageList from Symfony\Cmf\Bundle\SimpleCmsBundle\Tests\WebTest\Admin
```